Firestore/core/test/firebase/firestore/local/mutation_queue_test.cc

```diff
@@ -0,0 +1,432 @@
+/*
+ * Copyright 2019 Google
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "Firestore/core/test/firebase/firestore/local/mutation_queue_test.h"
+
+#include <memory>
+#include <utility>
+#include <vector>
+
+#include "Firestore/core/src/firebase/firestore/auth/user.h"
+#include "Firestore/core/src/firebase/firestore/local/persistence.h"
+#include "Firestore/core/src/firebase/firestore/model/document_key.h"
+#include "Firestore/core/src/firebase/firestore/model/document_key_set.h"
+#include "Firestore/core/src/firebase/firestore/model/mutation.h"
+#include "Firestore/core/src/firebase/firestore/model/mutation_batch.h"
+#include "Firestore/core/src/firebase/firestore/model/set_mutation.h"
+#include "Firestore/core/test/firebase/firestore/testutil/testutil.h"
+#include "gtest/gtest.h"
+
+namespace firebase {
+namespace firestore {
+namespace local {
+
+using auth::User;
+using model::DocumentKey;
+using model::DocumentKeySet;
+using model::kBatchIdUnknown;
+using model::Mutation;
+using model::MutationBatch;
+using model::SetMutation;
+using nanopb::ByteString;
+using testutil::Key;
+using testutil::Map;
+using testutil::Query;
+
+MutationQueueTestBase::MutationQueueTestBase(
+    std::unique_ptr<Persistence> persistence)
+    : persistence_(std::move(persistence)),
+      mutation_queue_(persistence_->GetMutationQueueForUser(User("user"))) {
+}
+
+MutationQueueTestBase::~MutationQueueTestBase() = default;
```

Comment:
Is there any advantage to not define it in the header?

Reply:
While `std::unique_ptr` is broadly compatible with forward declared classes, it's destructor is not. Moving the destructor of this class here after including the header that declares `Persistence` allows the header to mostly forward declare its dependencies.